### ISSUE TITLE
Fail to configure k8s_sat using use_token_review_api_validation

### ISSUE THREAD

**Issue by maxh8086** [spiffe/spire#956]
Hello Team,

Spire server isn't coming up whenever we configure use_token_review_api_validation under configmap.

https://github.com/spiffe/spire/blob/master/doc/plugin_server_nodeattestor_k8s_sat.md

```
kubectl describe cm -n spire spire-server
Name:         spire-server
Namespace:    spire
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"v1","data":{"bootstrap.crt":"-----BEGIN CERTIFICATE-----\nMIIBzDCCAVOgAwIBAgIJAJM4DhRH0vmuMAoGCCqGSM49BAMEMB4xCzAJBgNVBAYT\nAlVTMQ8wDQYD...

Data
====
bootstrap.crt:
----
-----BEGIN CERTIFICATE-----
MIIBzDCCAVOgAwIBAgIJAJM4DhRH0vmuMAoGCCqGSM49BAMEMB4xCzAJBgNVBAYT
AlVTMQ8wDQYDVQQKDAZTUElGRkUwHhcNMTgwNTEzMTkzMzQ3WhcNMjMwNTEyMTkz
MzQ3WjAeMQswCQYDVQQGEwJVUzEPMA0GA1UECgwGU1BJRkZFMHYwEAYHKoZIzj0C
AQYFK4EEACIDYgAEWjB+nSGSxIYiznb84xu5WGDZj80nL7W1c3zf48Why0ma7Y7m
CBKzfQkrgDguI4j0Z+0/tDH/r8gtOtLLrIpuMwWHoe4vbVBFte1vj6Xt6WeE8lXw
cCvLs/mcmvPqVK9jo10wWzAdBgNVHQ4EFgQUh6XzV6LwNazA+GTEVOdu07o5yOgw
DwYDVR0TAQH/BAUwAwEB/zAOBgNVHQ8BAf8EBAMCAQYwGQYDVR0RBBIwEIYOc3Bp
ZmZlOi8vbG9jYWwwCgYIKoZIzj0EAwQDZwAwZAIwE4Me13qMC9i6Fkx0h26y09QZ
IbuRqA9puLg9AeeAAyo5tBzRl1YL0KNEp02VKSYJAjBdeJvqjJ9wW55OGj1JQwDF
D7kWeEB6oMlwPbI/5hEY3azJi16I0uN1JSYTSWGSqWc=
-----END CERTIFICATE-----

server.conf:
----
server {
  bind_address = "0.0.0.0"
  bind_port = "8081"
  trust_domain = "example.org"
  data_dir = "/run/spire/data"
  log_level = "DEBUG"
  svid_ttl = "1h"
  ca_subject = {
    Country = ["IN"],
    Organization = ["SPIFFE"],
    CommonName = "",
  }
}

plugins {
  DataStore "sql" {
    plugin_data {
      database_type = "sqlite3"
      connection_string = "/run/spire/data/datastore.sqlite3"
    }
  }

  NodeAttestor "k8s_sat" {
    plugin_data {
      clusters = {
        # NOTE: Change this to your cluster name
        "k8s" = {
          # service_account_key_file = "/run/k8s-certs/sa.pub"
          service_account_whitelist = ["spire:spire-agent"]
          use_token_review_api_validation = true
        }
      }
    }
  }

  NodeResolver "noop" {
    plugin_data {}
  }

  KeyManager "disk" {
    plugin_data {
      keys_path = "/run/spire/data/keys.json"
    }
  }

  UpstreamCA "disk" {
    plugin_data {
      ttl = "12h"
      key_file_path = "/run/spire/secrets/bootstrap.key"
      cert_file_path = "/run/spire/config/bootstrap.crt"
    }
  }
}

Events:  <none>
[root@JMNGD1BAC10C07 k8s]# kubectl edit cm -n spire spire-server
configmap/spire-server edited
[root@JMNGD1BAC10C07 k8s]# kubectl logs -n spire spire-server-0
time="2019-06-11T12:36:42Z" level=warning msg="Current umask 0022 is too permissive; setting umask 0027."
time="2019-06-11T12:36:42Z" level=info msg="data directory: \"/run/spire/data\""
2019/06/11 12:36:42 [DEBUG] plugin: waiting for all plugin processes to complete...
time="2019-06-11T12:36:42Z" level=info msg="Starting plugin catalog" subsystem_name=catalog
time="2019-06-11T12:36:42Z" level=debug msg="NodeAttestor(k8s_sat): configuring plugin" subsystem_name=catalog
time="2019-06-11T12:36:42Z" level=info msg="Stopping plugin catalog" subsystem_name=catalog
time="2019-06-11T12:36:42Z" level=error msg="fatal: NodeAttestor(k8s_sat): failed to configure plugin: k8s-sat: cluster \"k8s\" configuration missing service account key file"
time="2019-06-11T12:36:42Z" level=error msg="NodeAttestor(k8s_sat): failed to configure plugin: k8s-sat: cluster \"k8s\" configuration missing service account key file"

```
* **Version**: <Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.1", GitCommit:"b1b29978270dc22fecc592ac55d903350454310a", GitTreeState:"clean", BuildDate:"2018-07-17T18:53:20Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.1", GitCommit:"b1b29978270dc22fecc592ac55d903350454310a", GitTreeState:"clean", BuildDate:"2018-07-17T18:43:26Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
>
* **Platform**: <CentOS>
* **Subsystem**: 
------------------------------

if we try to configure service account token with a key file, server run properly but failed to issue attach workload with SVID.

we are able to generate SVID from server but we are unable to fetch the same from client container.

---------------------
```
/opt/spire/bin # ./spire-server entry show
Found 2 entries
Entry ID      : b58c36e6-1a85-4ceb-bbb2-8dbff0260048
SPIFFE ID     : spiffe://example.org/client
Parent ID     : spiffe://example.org/spire/server
TTL           : 3600
Selector      : k8s:sa:default

Entry ID      : de6f160e-140d-4c95-9c8a-a0e2dd7283f8
SPIFFE ID     : spiffe://example.org/cluster
Parent ID     : spiffe://example.org/spire/server
TTL           : 3600
Selector      : k8s_sat:cluster:k8s

/opt/spire/bin # ./spire-server agent list
Found 3 attested agents:

Spiffe ID         : spiffe://example.org/spire/agent/k8s_sat/k8s/210dde30-82f8-45fe-bd17-a044b9a753fc
Attestation type  : k8s_sat
Expiration time   : 2019-06-11 13:47:02 +0000 UTC
Serial number     : 5

Spiffe ID         : spiffe://example.org/spire/agent/k8s_sat/k8s/d04778ff-db4d-4976-a31f-110ae0535544
Attestation type  : k8s_sat
Expiration time   : 2019-06-11 13:47:03 +0000 UTC
Serial number     : 6

Spiffe ID         : spiffe://example.org/spire/agent/k8s_sat/k8s/ea3a4f16-8520-4890-9740-fb23a12528f7
Attestation type  : k8s_sat
Expiration time   : 2019-06-11 13:47:05 +0000 UTC
Serial number     : 7

```--------------------------------

```
# /opt/spire/bin/spire-agent api watch -socketPath=/run/spire/sockets/agent.sock
2019/06/11 12:47:45 Received error from X509-SVID stream: rpc error: code = PermissionDenied desc = no identity issued
2019/06/11 12:47:46 Received error from X509-SVID stream: rpc error: code = PermissionDenied desc = no identity issued
2019/06/11 12:47:48 Received error from X509-SVID stream: rpc error: code = PermissionDenied desc = no identity issued
2019/06/11 12:47:52 Received error from X509-SVID stream: rpc error: code = PermissionDenied desc = no identity issued
2019/06/11 12:48:00 Received error from X509-SVID stream: rpc error: code = PermissionDenied desc = no identity issued
2019/06/11 12:48:16 Received error from X509-SVID stream: rpc error: code = PermissionDenied desc = no identity issued

```


agent config map :

```
kubectl get cm -n spire spire-agent
NAME          DATA      AGE
spire-agent   2         1d
[root@JMNGD1BAC10C07 k8s]# kubectl describe cm -n spire spire-agent
Name:         spire-agent
Namespace:    spire
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"v1","data":{"agent.conf":"agent {\n  data_dir = \"/run/spire\"\n  log_level = \"DEBUG\"\n  server_address = \"spire-server\"\n  server_p...

Data
====
agent.conf:
----
agent {
  data_dir = "/run/spire"
  log_level = "DEBUG"
  server_address = "spire-server"
  server_port = "8081"
  socket_path = "/run/spire/sockets/agent.sock"
  trust_bundle_path = "/run/spire/config/bootstrap.crt"
  trust_domain = "example.org"
}

plugins {
  NodeAttestor "k8s_sat" {
    plugin_data {
      # NOTE: Change this to your cluster name
      cluster = "k8s"
      # token_path = "/run/spire/tokens/spire-agent"
    }
  }

  KeyManager "memory" {
    plugin_data {
    }
  }

  WorkloadAttestor "k8s" {
    plugin_data {
      kubelet_read_only_port = "10255"
    }
  }

  WorkloadAttestor "unix" {
      plugin_data {
      }
  }
}

bootstrap.crt:
----
-----BEGIN CERTIFICATE-----
MIIBzDCCAVOgAwIBAgIJAJM4DhRH0vmuMAoGCCqGSM49BAMEMB4xCzAJBgNVBAYT
AlVTMQ8wDQYDVQQKDAZTUElGRkUwHhcNMTgwNTEzMTkzMzQ3WhcNMjMwNTEyMTkz
MzQ3WjAeMQswCQYDVQQGEwJVUzEPMA0GA1UECgwGU1BJRkZFMHYwEAYHKoZIzj0C
AQYFK4EEACIDYgAEWjB+nSGSxIYiznb84xu5WGDZj80nL7W1c3zf48Why0ma7Y7m
CBKzfQkrgDguI4j0Z+0/tDH/r8gtOtLLrIpuMwWHoe4vbVBFte1vj6Xt6WeE8lXw
cCvLs/mcmvPqVK9jo10wWzAdBgNVHQ4EFgQUh6XzV6LwNazA+GTEVOdu07o5yOgw
DwYDVR0TAQH/BAUwAwEB/zAOBgNVHQ8BAf8EBAMCAQYwGQYDVR0RBBIwEIYOc3Bp
ZmZlOi8vbG9jYWwwCgYIKoZIzj0EAwQDZwAwZAIwE4Me13qMC9i6Fkx0h26y09QZ
IbuRqA9puLg9AeeAAyo5tBzRl1YL0KNEp02VKSYJAjBdeJvqjJ9wW55OGj1JQwDF
D7kWeEB6oMlwPbI/5hEY3azJi16I0uN1JSYTSWGSqWc=
-----END CERTIFICATE-----

```



**logs from spire agent -** 

--------------
time="2019-06-11T12:46:55Z" level=error msg="synchronize failed: unable to get a stream" subsystem_name=manager
time="2019-06-11T12:46:59Z" level=debug msg="cache entry ttl for spiffeId spiffe://example.org/cluster is less than a half its lifetime" subsystem_name=manager
time="2019-06-11T12:46:59Z" level=debug msg="Requesting SVID for spiffe://example.org/cluster" subsystem_name=manager
time="2019-06-11T12:47:08Z" level=debug msg="Rotating agent SVID" subsystem_name=manager
time="2019-06-11T12:47:54Z" level=error msg="Failed to collect all selectors for PID 126520: workload attestor \"k8s\" failed: k8s: Get http://localhost:10255/pods: dial tcp 165.160.13.20:10255: i/o timeout" subsystem_name=workload_api
time="2019-06-11T12:47:54Z" level=debug msg="PID 126520 attested to have selectors [type:\"unix\" value:\"uid:0\"  type:\"unix\" value:\"user:root\"  type:\"unix\" value:\"gid:0\"  type:\"unix\" value:\"group:root\" ]" subsystem_name=workload_api
time="2019-06-11T12:48:54Z" level=error msg="Failed to collect all selectors for PID 127081: workload attestor \"k8s\" failed: k8s: Get http://localhost:10255/pods: dial tcp 165.160.15.20:10255: i/o timeout" subsystem_name=workload_api
time="2019-06-11T12:48:54Z" level=debug msg="PID 127081 attested to have selectors [type:\"unix\" value:\"uid:0\"  type:\"unix\" value:\"user:root\"  type:\"unix\" value:\"gid:0\"  type:\"unix\" value:\"group:root\" ]" subsystem_name=workload_api
time="2019-06-11T12:49:54Z" level=error msg="Failed to collect all selectors for PID 127843: workload attestor \"k8s\" failed: k8s: Get http://localhost:10255/pods: dial tcp 165.160.13.20:10255: i/o timeout" subsystem_name=workload_api
time="2019-06-11T12:49:54Z" level=debug msg="PID 127843 attested to have selectors [type:\"unix\" value:\"uid:0\"  type:\"unix\" value:\"user:root\"  type:\"unix\" value:\"gid:0\"  type:\"unix\" value:\"group:root\" ]" subsystem_name=workload_api
time="2019-06-11T12:50:54Z" level=error msg="Failed to collect all selectors for PID 128743: workload attestor \"k8s\" failed: k8s: Get http://localhost:10255/pods: dial tcp 165.160.15.20:10255: i/o timeout" subsystem_name=workload_api
time="2019-06-11T12:50:54Z" level=debug msg="PID 128743 attested to have selectors [type:\"unix\" value:\"uid:0\"  type:\"unix\" value:\"user:root\"  type:\"unix\" value:\"gid:0\"  type:\"unix\" value:\"group:root\" ]" subsystem_name=workload_api

**Comment by calaniz** [comment]
The kubelet is likely listening on 10250.  This is the secure communication port on recent clusters. In kubeadm spun up clusters, the certificate for this is self signed.  You'll also need to add skip_kubelet_verification to true in your agent conf.

Both of these settings apply to the k8s workload attestor for the agent.

**Comment by maxh8086** [comment]
i changed port to 10250, didn't work, changed it back to 10255 as of now.
updated skip_kubelet_verification under workload attestor under agent cm. still getting below errors.

```
kubectl logs -n spire spire-agent-hjv99
time="2019-06-11T14:33:04Z" level=warning msg="Current umask 0022 is too permissive; setting umask 0027."
time="2019-06-11T14:33:04Z" level=info msg="data directory: \"/run/spire\""
time="2019-06-11T14:33:04Z" level=info msg="Starting plugin catalog" subsystem_name=catalog
time="2019-06-11T14:33:04Z" level=debug msg="NodeAttestor(k8s_sat): configuring plugin" subsystem_name=catalog
time="2019-06-11T14:33:04Z" level=debug msg="KeyManager(memory): configuring plugin" subsystem_name=catalog
time="2019-06-11T14:33:04Z" level=debug msg="WorkloadAttestor(unix): configuring plugin" subsystem_name=catalog
time="2019-06-11T14:33:04Z" level=debug msg="WorkloadAttestor(k8s): configuring plugin" subsystem_name=catalog
time="2019-06-11T14:33:04Z" level=debug msg="No pre-existing agent SVID found. Will perform node attestation" subsystem_name=attestor
time="2019-06-11T14:33:04Z" level=debug msg="Requesting SVID for spiffe://example.org/cluster" subsystem_name=manager
time="2019-06-11T14:33:04Z" level=info msg="Starting workload API" subsystem_name=endpoints
time="2019-06-11T14:34:38Z" level=error msg="Failed to collect all selectors for PID 70017: workload attestor \"k8s\" failed: k8s: Get http://localhost:10255/pods: dial tcp 165.160.15.20:10255: i/o timeout" subsystem_name=workload_api
time="2019-06-11T14:34:38Z" level=debug msg="PID 70017 attested to have selectors [type:\"unix\" value:\"uid:0\"  type:\"unix\" value:\"user:root\"  type:\"unix\" value:\"gid:0\"  type:\"unix\" value:\"group:root\" ]" subsystem_name=workload_api

```

**Comment by evan2645** [comment]
> time="2019-06-11T12:36:42Z" level=error msg="fatal: NodeAttestor(k8s_sat): failed to configure plugin: k8s-sat: cluster \"k8s\" configuration missing service account key file"
> time="2019-06-11T12:36:42Z" level=error msg="NodeAttestor(k8s_sat): failed to configure plugin: k8s-sat: cluster \"k8s\" configuration missing service account key file"

This error message should only occur when `use_token_review_api_validation` is not set. In the logs you pasted it shows an edit to the config map - can you verify that spire-server received the config map update? The config you pasted looks correct...

```
Entry ID      : b58c36e6-1a85-4ceb-bbb2-8dbff0260048
SPIFFE ID     : spiffe://example.org/client
Parent ID     : spiffe://example.org/spire/server
TTL           : 3600
Selector      : k8s:sa:default
```

Parent ID should be `spiffe://example.org/cluster` and not spire server. In plain english, this says "when the workload is running in spiffe://example.org/cluster, and in k8s sa default, issue spiffe id spiffe://example.org/client". Incorrect parent ID will definitely cause you to see no identity issued. Once the parent id is fixed/updated, you should see in SPIRE server logs that SVID for that SPIFFE ID is issued, and in agent logs that it has cached the certificate

**Comment by evan2645** [comment]
> time="2019-06-11T14:34:38Z" level=error msg="Failed to collect all selectors for PID 70017: workload attestor \"k8s\" failed: k8s: Get http://localhost:10255/pods: dial tcp 165.160.15.20:10255: i/o timeout" subsystem_name=workload_api

This also looks like a problem, and we can see in the following log line that the agent is not discovering the k8s selectors as a result. Agent is timing out trying to talk to kubelet. Make sure that the agent is in host network namespace so that it can actually reach the kubelet port: https://github.com/spiffe/spire/blob/master/examples/k8s/simple_sat/spire-agent.yaml#L82-L84

**Comment by maxh8086** [comment]
Verified agent ds, setting is present : 

      dnsPolicy: ClusterFirstWithHostNet
      hostNetwork: true
      hostPID: true

changed port to 10250 post verifying netstat -n on worker / master node, found 10250 is in listening mode.
Present status of agent.conf from inside pod.

```
kubectl exec -it spire-agent-6rh6d sh -n spire
/opt/spire # cat /run/spire/config/agent.conf
agent {
  data_dir = "/run/spire"
  log_level = "DEBUG"
  server_address = "spire-server"
  server_port = "8081"
  socket_path = "/run/spire/sockets/agent.sock"
  trust_bundle_path = "/run/spire/config/bootstrap.crt"
  trust_domain = "example.org"
}

plugins {
  NodeAttestor "k8s_sat" {
    plugin_data {
      # NOTE: Change this to your cluster name
      cluster = "k8s"
      # token_path = "/run/spire/tokens/spire-agent"
    }
  }

  KeyManager "memory" {
    plugin_data {
    }
  }

  WorkloadAttestor "k8s" {
    plugin_data {
      kubelet_read_only_port = "10250"
      skip_kubelet_verification = true
    }
  }

  WorkloadAttestor "unix" {
      plugin_data {
      }
  }
}

```
Present logs of agent pod -

```
 kubectl logs -n spire spire-agent-6rh6d | more
time="2019-06-12T14:31:40Z" level=warning msg="Current umask 0022 is too permissive; setting umask 0027."
time="2019-06-12T14:31:40Z" level=info msg="data directory: \"/run/spire\""
time="2019-06-12T14:31:40Z" level=info msg="Starting plugin catalog" subsystem_name=catalog
time="2019-06-12T14:31:40Z" level=debug msg="WorkloadAttestor(k8s): configuring plugin" subsystem_name=catalog
time="2019-06-12T14:31:40Z" level=debug msg="WorkloadAttestor(unix): configuring plugin" subsystem_name=catalog
time="2019-06-12T14:31:40Z" level=debug msg="NodeAttestor(k8s_sat): configuring plugin" subsystem_name=catalog
time="2019-06-12T14:31:40Z" level=debug msg="KeyManager(memory): configuring plugin" subsystem_name=catalog
time="2019-06-12T14:31:40Z" level=debug msg="No pre-existing agent SVID found. Will perform node attestation" subsystem_name=attestor
time="2019-06-12T14:31:40Z" level=debug msg="Requesting SVID for spiffe://example.org/cluster" subsystem_name=manager
time="2019-06-12T14:31:40Z" level=info msg="Starting workload API" subsystem_name=endpoints
time="2019-06-12T14:32:11Z" level=error msg="Failed to collect all selectors for PID 72678: workload attestor \"k8s\" failed: k8s: Get http://localhost:10250/pods: net
/http: HTTP/1.x transport connection broken: malformed HTTP response \"\\x15\\x03\\x01\\x00\\x02\\x02\"" subsystem_name=workload_api
time="2019-06-12T14:32:11Z" level=debug msg="PID 72678 attested to have selectors [type:\"unix\" value:\"uid:0\"  type:\"unix\" value:\"user:root\"  type:\"unix\" valu
e:\"gid:0\"  type:\"unix\" value:\"group:root\" ]" subsystem_name=workload_api
time="2019-06-12T14:33:11Z" level=error msg="Failed to collect all selectors for PID 76193: workload attestor \"k8s\" failed: k8s: Get http://localhost:10250/pods: net
/http: HTTP/1.x transport connection broken: malformed HTTP response \"\\x15\\x03\\x01\\x00\\x02\\x02\"" subsystem_name=workload_api
time="2019-06-12T14:33:11Z" level=debug msg="PID 76193 attested to have selectors [type:\"unix\" value:\"uid:0\"  type:\"unix\" value:\"user:root\"  type:\"unix\" valu
e:\"gid:0\"  type:\"unix\" value:\"group:root\" ]" subsystem_name=workload_api
time="2019-06-12T14:34:11Z" level=error msg="Failed to collect all selectors for PID 79648: workload attestor \"k8s\" failed: k8s: Get http://localhost:10250/pods: net
/http: HTTP/1.x transport connection broken: malformed HTTP response \"\\x15\\x03\\x01\\x00\\x02\\x02\"" subsystem_name=workload_api
time="2019-06-12T14:34:11Z" level=debug msg="PID 79648 attested to have selectors [type:\"unix\" value:\"uid:0\"  type:\"unix\" value:\"user:root\"  type:\"unix\" valu
e:\"gid:0\"  type:\"unix\" value:\"group:root\" ]" subsystem_name=workload_api
time="2019-06-12T14:35:11Z" level=error msg="Failed to collect all selectors for PID 83208: workload attestor \"k8s\" failed: k8s: Get http://localhost:10250/pods: net
/http: HTTP/1.x transport connection broken: malformed HTTP response \"\\x15\\x03\\x01\\x00\\x02\\x02\"" subsystem_name=workload_api
time="2019-06-12T14:35:11Z" level=debug msg="PID 83208 attested to have selectors [type:\"unix\" value:\"uid:0\"  type:\"unix\" value:\"user:root\"  type:\"unix\" valu
e:\"gid:0\"  type:\"unix\" value:\"group:root\" ]" subsystem_name=workload_api

```

**Comment by azdagron** [comment]
You have set the kubelet_read_only_port option to 10250 (secure kubelet port). This directs the attestor to attempt to talk the the kubelet via HTTP instead of HTTPS. If you remove that option completely, it will use the default secure kubelet port over HTTPS.

**Comment by maxh8086** [comment]
i removed port option completely from configmap now but got below error : 

```
kubectl logs -n spire spire-agent-5dk6z
time="2019-06-14T07:39:46Z" level=warning msg="Current umask 0022 is too permissive; setting umask 0027."
time="2019-06-14T07:39:46Z" level=info msg="data directory: \"/run/spire\""
time="2019-06-14T07:39:46Z" level=info msg="Starting plugin catalog" subsystem_name=catalog
time="2019-06-14T07:39:46Z" level=debug msg="NodeAttestor(k8s_sat): configuring plugin" subsystem_name=catalog
time="2019-06-14T07:39:46Z" level=debug msg="KeyManager(memory): configuring plugin" subsystem_name=catalog
time="2019-06-14T07:39:46Z" level=debug msg="WorkloadAttestor(k8s): configuring plugin" subsystem_name=catalog
time="2019-06-14T07:39:46Z" level=debug msg="WorkloadAttestor(unix): configuring plugin" subsystem_name=catalog
time="2019-06-14T07:39:46Z" level=debug msg="No pre-existing agent SVID found. Will perform node attestation" subsystem_name=attestor
time="2019-06-14T07:39:46Z" level=debug msg="Requesting SVID for spiffe://example.org/cluster" subsystem_name=manager
time="2019-06-14T07:39:46Z" level=info msg="Starting workload API" subsystem_name=endpoints
time="2019-06-14T07:40:44Z" level=error msg="Failed to collect all selectors for PID 4101: workload attestor \"k8s\" failed: k8s: Get http://localhost:0/pods: dial tcp [::1]:0: connect: connection refused" subsystem_name=workload_api
time="2019-06-14T07:40:44Z" level=debug msg="PID 4101 attested to have selectors [type:\"unix\" value:\"uid:0\"  type:\"unix\" value:\"user:root\"  type:\"unix\" value:\"gid:0\"  type:\"unix\" value:\"group:root\" ]" subsystem_name=workload_api

``` 


**time="2019-06-14T07:40:44Z" level=error msg="Failed to collect all selectors for PID 4101: workload attestor \"k8s\" failed: k8s: Get http://localhost:0/pods: dial tcp [::1]:0: connect: connection refused" subsystem_name=workload_api**

error appears again here, i changed port to 10251 now  and got below logs, where its now not attaching workload but above error disappeared.

Logs after making change :  

```
nginx-ingress]# kubectl logs -n spire spire-agent-25zff
time="2019-06-14T09:27:26Z" level=warning msg="Current umask 0022 is too permissive; setting umask 0027."
time="2019-06-14T09:27:26Z" level=info msg="data directory: \"/run/spire\""
time="2019-06-14T09:27:26Z" level=info msg="Starting plugin catalog" subsystem_name=catalog
time="2019-06-14T09:27:26Z" level=debug msg="NodeAttestor(k8s_sat): configuring plugin" subsystem_name=catalog
time="2019-06-14T09:27:26Z" level=debug msg="KeyManager(memory): configuring plugin" subsystem_name=catalog
time="2019-06-14T09:27:26Z" level=debug msg="WorkloadAttestor(k8s): configuring plugin" subsystem_name=catalog
time="2019-06-14T09:27:26Z" level=debug msg="WorkloadAttestor(unix): configuring plugin" subsystem_name=catalog
time="2019-06-14T09:27:26Z" level=debug msg="No pre-existing agent SVID found. Will perform node attestation" subsystem_name=attestor
time="2019-06-14T09:27:26Z" level=debug msg="Requesting SVID for spiffe://example.org/cluster" subsystem_name=manager
time="2019-06-14T09:27:26Z" level=info msg="Starting workload API" subsystem_name=endpoints
time="2019-06-14T09:28:20Z" level=error msg="Failed to collect all selectors for PID 3609: workload attestor \"k8s\" failed: k8s: unexpected status code: 404" subsystem_name=workload_api
time="2019-06-14T09:28:20Z" level=debug msg="PID 3609 attested to have selectors [type:\"unix\" value:\"uid:0\"  type:\"unix\" value:\"user:root\"  type:\"unix\" value:\"gid:0\"  type:\"unix\" value:\"group:root\" ]" subsystem_name=workload_api
time="2019-06-14T09:29:20Z" level=error msg="Failed to collect all selectors for PID 7448: workload attestor \"k8s\" failed: k8s: unexpected status code: 404" subsystem_name=workload_api
time="2019-06-14T09:29:20Z" level=debug msg="PID 7448 attested to have selectors [type:\"unix\" value:\"uid:0\"  type:\"unix\" value:\"user:root\"  type:\"unix\" value:\"gid:0\"  type:\"unix\" value:\"group:root\" ]" subsystem_name=workload_api
time="2019-06-14T09:30:20Z" level=error msg="Failed to collect all selectors for PID 11406: workload attestor \"k8s\" failed: k8s: unexpected status code: 404" subsystem_name=workload_api
time="2019-06-14T09:30:20Z" level=debug msg="PID 11406 attested to have selectors [type:\"unix\" value:\"uid:0\"  type:\"unix\" value:\"user:root\"  type:\"unix\" value:\"gid:0\"  type:\"unix\" value:\"group:root\" ]" subsystem_name=workload_api
time="2019-06-14T09:31:20Z" level=error msg="Failed to collect all selectors for PID 15080: workload attestor \"k8s\" failed: k8s: unexpected status code: 404" subsystem_name=workload_api

```


Below is netstat -n | grep -i 1025 result of my server (changed actual IP's)

```
tcp        0      0 1.1.97.19:54018     1.1.113.15:10250    ESTABLISHED
tcp        0      0 127.0.0.1:10251         127.0.0.1:48964         TIME_WAIT
tcp        0      0 127.0.0.1:40650         127.0.0.1:10252         TIME_WAIT
tcp        0      0 1.1.97.19:45996     1.1.21.18:10250     ESTABLISHED
tcp        0      0 127.0.0.1:10252         127.0.0.1:40456         TIME_WAIT
tcp        0      0 127.0.0.1:10251         127.0.0.1:48818         TIME_WAIT
tcp        0      0 127.0.0.1:10251         127.0.0.1:48770         TIME_WAIT
tcp        0      0 127.0.0.1:10252         127.0.0.1:40550         TIME_WAIT
tcp        0      0 1.1.97.19:40052     1.1.113.13:10250    ESTABLISHED
tcp        0      0 1.1.97.19:53688     1.1.97.19:10250     ESTABLISHED
tcp        0      0 1.1.97.19:51918     1.1.97.19:10250     ESTABLISHED
tcp        0      0 1.1.97.19:40050     1.1.113.13:10250    ESTABLISHED
tcp        0      0 1.1.97.19:53230     1.1.97.19:10250     ESTABLISHED
tcp        0      0 1.1.97.19:44906     1.1.113.13:10250    ESTABLISHED
tcp        0      0 1.1.97.19:45926     1.1.21.18:10250     ESTABLISHED
tcp        0      0 127.0.0.1:10252         127.0.0.1:40502         TIME_WAIT
tcp        0      0 127.0.0.1:10252         127.0.0.1:40600         TIME_WAIT
tcp        0      0 127.0.0.1:10251         127.0.0.1:48918         TIME_WAIT
tcp        0      0 1.1.97.19:47250     1.1.97.19:10250     ESTABLISHED
tcp        0      0 1.1.97.19:45828     1.1.21.18:10250     ESTABLISHED
tcp        0      0 127.0.0.1:10252         127.0.0.1:40700         TIME_WAIT
tcp        0      0 127.0.0.1:10251         127.0.0.1:48864         TIME_WAIT
tcp        0      0 127.0.0.1:10252         127.0.0.1:40650         TIME_WAIT
tcp        0      0 127.0.0.1:48906         127.0.0.1:10251         TIME_WAIT
tcp6       0      0 1.1.97.19:10250     1.1.97.19:51918     ESTABLISHED
tcp6       0      0 1.1.97.19:10250     1.1.97.19:53688     ESTABLISHED
tcp6       0      0 1.1.97.19:10250     10.32.1.177:57362       ESTABLISHED
tcp6       0      0 1.1.97.19:10250     1.1.97.19:47250     ESTABLISHED
tcp6       0      0 1.1.97.19:10250     1.1.97.19:53230     ESTABLISHED

```

**Comment by marcosy** [comment]
> Fail to configure k8s_sat using use_token_review_api_validation

Hello! Which SPIRE version are you using? Token review API validation for SAT attestor is available starting on SPIRE 0.8.0. 
You should also set `upstream_bundle = true` in the `server { ... }` section of your server.conf file when using an upstreamCA bootstrap certificate in SPIRE 0.8.0.

**Comment by evan2645** [comment]
```
k8s: unexpected status code: 404
```

I have dug through the codebase and found that this error string does not exist in recent versions of SPIRE. It was changed on March 21 with the same commit that added support for the secure kubelet port. So:
* The configuration in which you received error message `malformed HTTP response \"\\x15\\x03\\x01\\x00\\x02\\x02\""` was likely correct
* Your kubelet is serving the secure port, but the version you are running does not support the secure port
* The version you are running doesn't support using the token review API either, which is why it appeared that the configuration option was being ignored
* Upgrade to SPIRE 0.8.0 and follow the secure kubelet configuration guidance in the documentation, that should hopefully work for you

**Comment by evan2645** [comment]
@maxh8086 Were you ever able to get this resolved?

**Comment by evan2645** [comment]
@maxh8086 closing this out since it appears to have been a misconfiguration, and we have not heard back. Please drop another comment here if you're still experiencing problems and we can open it back up. Alternatively, I recommend you join our Slack organization for faster support https://slack.spiffe.io
